packages/salesforcedx-vscode-core/src/deploys/toolingDeploy.ts

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { Connection } from '@salesforce/core';
+import * as fs from 'fs';
+import * as path from 'path';
+import { nls } from '../messages';
+import {
+  DeployStatusEnum,
+  supportedToolingTypes,
+  ToolingCreateResult,
+  ToolingRetrieveResult
+} from './index';
+// tslint:disable-next-line:no-var-requires
+const DOMParser = require('xmldom-sfdx-encoding').DOMParser;
+const CONTAINER_ASYNC_REQUEST = 'ContainerAsyncRequest';
+const METADATA_CONTAINER = 'MetadataContainer';
+
+export class ToolingDeploy {
+  public metadataType?: string;
+  public connection: Connection;
+  private apiVersion?: string;
+
+  public constructor(connection: Connection, apiVersion?: string) {
+    this.connection = connection;
+    if (apiVersion) this.apiVersion = apiVersion;
+  }
+
+  public async deploy(filePath: string) {
+    // find out what the type is from metadataregistry given the sourcepath and assign
+    this.metadataType = 'Apexclass';
+    const sourcePath = filePath.replace('-meta.xml', '');
+    const metadataPath = `${sourcePath}-meta.xml`;
+
+    const container = await this.createMetadataContainer();
+    await this.createContainerMember([sourcePath, metadataPath], container);
+    const asyncRequest = await this.createContainerAsyncRequest(container);
+    const output = await this.toolingRetrieve(asyncRequest);
+    return output;
+  }
+
+  public async createMetadataContainer(): Promise<ToolingCreateResult> {
+    const metadataContainer = (await this.connection.tooling.create(
+      METADATA_CONTAINER,
+      { Name: `VSCode_MDC_${Date.now()}` }
+    )) as ToolingCreateResult;
+
+    if (!metadataContainer.success) {
+      const deployFailed = new Error();
+      deployFailed.message = nls.localize('beta_tapi_mdcontainer_error');
+      deployFailed.name = 'MetadataContainerCreationFailed';
+      throw deployFailed;
+    }
+    return metadataContainer;
+  }
+
+  public buildMetadataField(metadataContent: string) {
+    const parser = new DOMParser();
+    const document = parser.parseFromString(metadataContent, 'text/xml');
+    const apiVersion =
```

Comment:
Is it possible for these values to be different? And if so, would that cause any issues for us?

Reply:
Yes, normally we want to use the apiVersion on the `-meta.xml` files (document.getElementsByTagName('apiVersion')...) but there might be scenarios in which we want to override it and that's when we would define `this.apiVersion`. Most of the time we want to use the api version defined in the `-meta.xml` file since some features might behave differently in the org depending on their api version, specially when it comes to Apex where some features are deprecated or introduced in specific api versions.